python/private/pypi/whl_library_targets.bzl

```diff
@@ -0,0 +1,339 @@
+        )]
+
+        native.alias(
+            name = PY_LIBRARY_PUBLIC_LABEL,
+            actual = label_tmpl.format(PY_LIBRARY_PUBLIC_LABEL),
+            visibility = ["//visibility:public"],
+        )
+        native.alias(
+            name = WHEEL_FILE_PUBLIC_LABEL,
+            actual = label_tmpl.format(WHEEL_FILE_PUBLIC_LABEL),
+            visibility = ["//visibility:public"],
+        )
+        py_library_label = PY_LIBRARY_IMPL_LABEL
+        whl_file_label = WHEEL_FILE_IMPL_LABEL
+
+    elif group_name:
+        py_library_label = PY_LIBRARY_PUBLIC_LABEL
+        whl_file_label = WHEEL_FILE_PUBLIC_LABEL
+        impl_vis = [dep_template.format(name = "", target = "__subpackages__")]
+
+    else:
+        py_library_label = PY_LIBRARY_PUBLIC_LABEL
+        whl_file_label = WHEEL_FILE_PUBLIC_LABEL
+        impl_vis = ["//visibility:public"]
+
+    if hasattr(native, "filegroup"):
+        native.filegroup(
+            name = whl_file_label,
+            srcs = [whl_name],
+            data = _deps(
+                deps = dependencies,
+                deps_by_platform = dependencies_by_platform,
+                tmpl = dep_template.format(name = "{}", target = WHEEL_FILE_PUBLIC_LABEL),
+                select = getattr(native, "select", select),
+            ),
+            visibility = impl_vis,
+        )
+
+    if hasattr(rules, "py_library"):
+        _data_exclude = [
+            "**/* *",
+            "**/*.py",
+            "**/*.pyc",
+            "**/*.pyc.*",  # During pyc creation, temp files named *.pyc.NNNN are created
+            # RECORD is known to contain sha256 checksums of files which might include the checksums
+            # of generated files produced when wheels are installed. The file is ignored to avoid
+            # Bazel caching issues.
+            "**/*.dist-info/RECORD",
+        ]
+        for item in data_exclude:
+            if item not in _data_exclude:
+                _data_exclude.append(item)
+
+        rules.py_library(
+            name = py_library_label,
+            srcs = native.glob(
+                ["site-packages/**/*.py"],
+                exclude = srcs_exclude,
+                # Empty sources are allowed to support wheels that don't have any
+                # pure-Python code, e.g. pymssql, which is written in Cython.
+                allow_empty = True,
+            ),
+            data = data + native.glob(
+                ["site-packages/**/*"],
+                exclude = _data_exclude,
+            ),
+            # This makes this directory a top-level in the python import
+            # search path for anything that depends on this.
+            imports = ["site-packages"],
+            deps = _deps(
+                deps = dependencies,
+                deps_by_platform = dependencies_by_platform,
+                tmpl = dep_template.format(name = "{}", target = PY_LIBRARY_PUBLIC_LABEL),
+                select = getattr(native, "select", select),
```

Comment:
This is a weird getattr() ? When does native.select _not_ exist??

Reply:
```suggestion
                # Actually, `select` is not part of `native`, but in order to support
                # bazel 6.4 in unit tests, I have to somehow pass the `select` 
                # implementation in the unit tests and I chose this to be routed through
                # the `native` struct. So, tests` will be successful in `getattr` and
                # the real code will use the fallback provided here.
                select = getattr(native, "select", select),
```